airbyte-integrations/connectors/source-github/source_github/streams.py

```diff
@@ -79,94 +92,53 @@ def request_params(
-            if headers:
-                headers = f"HTTP headers: {headers},"
-
-            self.logger.info(
-                f"Rate limit handling for stream `{self.name}` for the response with {response.status_code} status code, {headers} with message: {response.text}"
-            )
-
-        return retry_flag
 
-    def backoff_time(self, response: requests.Response) -> Optional[float]:
-        # This method is called if we run into the rate limit. GitHub limits requests to 5000 per hour and provides
-        # `X-RateLimit-Reset` header which contains time when this hour will be finished and limits will be reset so
-        # we again could have 5000 per another hour.
-
-        min_backoff_time = 60.0
-        retry_after = response.headers.get("Retry-After")
-        if retry_after is not None:
-            backoff_time_in_seconds = max(float(retry_after), min_backoff_time)
-            return self.get_waiting_time(backoff_time_in_seconds)
-
-        reset_time = response.headers.get("X-RateLimit-Reset")
-        if reset_time:
-            backoff_time_in_seconds = max(float(reset_time) - time.time(), min_backoff_time)
-            return self.get_waiting_time(backoff_time_in_seconds)
-
-    def get_waiting_time(self, backoff_time_in_seconds):
-        if backoff_time_in_seconds < self.max_time:
-            return backoff_time_in_seconds
-        else:
-            self._session.auth.update_token()  # New token will be used in next request
-            return 1
+    def get_backoff_strategy(self) -> Optional[Union[BackoffStrategy, List[BackoffStrategy]]]:
+        return GithubStreamABCBackoffStrategy(stream=self)
 
     @staticmethod
-    def check_graphql_rate_limited(response_json: dict) -> bool:
+    def check_graphql_rate_limited(response_json: Dict[str, Any]) -> bool:
         errors = response_json.get("errors")
         if errors:
             for error in errors:
                 if error.get("type") == "RATE_LIMITED":
                     return True
         return False
 
-    def read_records(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: Optional[List[str]] = None,
+        stream_slice: Optional[Mapping[str, Any]] = None,
+        stream_state: Optional[Mapping[str, Any]] = None,
+    ) -> Iterable[Mapping[str, Any]]:
         # get out the stream_slice parts for later use.
         organisation = stream_slice.get("organization", "")
         repository = stream_slice.get("repository", "")
         # Reading records while handling the errors
         try:
-            yield from super().read_records(stream_slice=stream_slice, **kwargs)
-        except HTTPError as e:
+            yield from super().read_records(
+                sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state
+            )
+        except DefaultBackoffException as e:
```

Comment:
Can we remove this exception handling and move it to the backoff logic?

Reply:
What do you mean by backoff logic? If you are referring to an error handler, we cannot simply ignore an error. Even in case of an error, we still need to parse the response and this can be problematic.

Reply:
> If you are referring to an error handler, we cannot simply ignore an error

Yes, you're right, I was talking about `error_handler`. 

So, the only reason why we set `response_action=ResponseAction.RETRY` in `ErrorResolution` and catch `DefaultBackoffException` afterwards is the response parsing?

I'm asking, because my initial thought was to move all exception handlers to error_handler instance inside `HttpClient`.

Reply:
Yes, this is a big problem for us because when we ignore errors, we parse the response using the same rules as for a successful result. Therefore, we cannot move this custom error handling to the error handler.

Reply:
hmm, ok, then I'm good wit this for now.

We should consider adding callable fn(response) to `HttpStatusErrorHandler`, in order to generate error_messages dynamically (in `error_mapping`).